### PR TITLE
Fix compatibility issue with Harbor token service

### PR DESCRIPTION
**Pull request by l2dy**
When a request contains `Authorization` header but fails to authenticate, Harbor registry returns `Basic realm="harbor"` as the authentication challenge [1], instead of the token service configured.

nerdctl caches pull-scoped tokens for later use, so when a *push* request is sent with the cached pull-scoped token, Harbor denies the request and instructs the client to attempt "Basic" authentication, which consistently fails if "Basic" authentication is disabled or reserved for special accounts only.

`nerdctl push` to a Harbor registry could fall into this "Basic" authentication loop and eventually fail with 401 Unauthorized.

This patch detects such pattern of repeated requests and retries the request without `Authorization` header to get a proper "Bearer" challenge from the registry.

[1]: https://github.com/goharbor/harbor/blob/6113469a567623054a58d0f5506b90a55adca7ff/src/server/middleware/v2auth/auth.go#L95-L98

Fixes: https://github.com/containerd/nerdctl/issues/1675

### CODE REVIEW

**Comment by l2dy** [comment]
@fahedouch Should this fix be implemented in nerdctl or containerd's `remotes/docker/authorizer.go`? I will proceed testing this change after your decision.

**Review by fahedouch — CHANGES_REQUESTED** [review]

**Comment by fahedouch** [comment]
> @fahedouch Should this fix be implemented in nerdctl or containerd's `remotes/docker/authorizer.go`? I will proceed testing this change after your decision.

login is on nerdctl side. SGTM to keep it on nerdctl

File: pkg/cmd/login/login.go [hunk 1]
```diff
@@ -225,8 +225,12 @@ func tryLoginWithRegHost(ctx context.Context, rh docker.RegistryHost) error {
...
 			}
 		}
-		if err := rh.Authorizer.Authorize(ctx, req); err != nil {
-			return fmt.Errorf("failed to call rh.Authorizer.Authorize: %w", err)
+		// If stuck in a loop of the same unmet `WWW-Authenticate` challenge,
+		// send a request without `Authorization` header.
+		if !infiniteAuthorizationLoop(ress) {
+			if err := rh.Authorizer.Authorize(ctx, req); err != nil {
+				return fmt.Errorf("failed to call rh.Authorizer.Authorize: %w", err)
+			}
```
[the remark this answers is not in the document]
Comment (l2dy) [reply pkg/cmd/login/login.go:233]: > if **NOT** `infiniteAuthorizationLoop` => `Authorize`, So how to authorize when we have an `infiniteAuthorizationLoop`? .

Proper authorization should happen in the next iteration, if the authentication method that the server requests (in its response) in this iteration has changed.

A side effect is that if the server returns 401 and the same `WWW-Authenticate` header for a temporary failure, retry only works once. Later retries will not contain an `Authorization` Header at all.

I'm assuming that this package is only used in the command line, where state is not persisted across executions, so this side effect is negligible.

> as discussed int t[he related issue ticket](https://github.com/containerd/nerdctl/issues/1675), an `Authorization` Header within an authorization request is the cause of the infiniteLoop when targeting harbor registry, So we have to remove this header to break the infiniteLoop and get a valid `token`. I believe that no need for an authorization header when making an Authorization request. So we can, by default, ensure that Authorization request is proceeded without `Authorization` Header, by removing the `Authorization` Header from `req` before `rh.Authorizer.Authorize(ctx, req)`. So, then drop the infiniteLoop check mechanisms

`rh.Authorizer.Authorize(ctx, req)` sets the `Authorization` Header, so clearing it before the statement is no-op.

File: pkg/cmd/login/login.go [hunk 2]
```diff
@@ -321,3 +325,28 @@ func convertToHostname(serverAddress string) (string, error) {
...
 }
+
+func infiniteAuthorizationLoop(responses []*http.Response) bool {
+	n := len(responses)
+	if n == 1 || (n > 1 && !sameUnmetAuthorizationChallenge(responses[n-2], responses[n-1])) {
+		return false
+	}
+
+	return true
+}
```
Comment (AkihiroSuda) [inline pkg/cmd/login/login.go:336]: Hard to understand this function

Comment (fahedouch) [inline pkg/cmd/login/login.go:336]: When `n = 0` you return `true`. Is it normal ?
If `n = 0` return false => no need for next iteration to make a propre Authorization

Comment (l2dy) [reply pkg/cmd/login/login.go:336]: It's not, thanks for spotting this. I copied the logic from https://github.com/containerd/containerd/blob/f0ce2f6d369a35d9b0bf3e6a31404ee3dbfdb9bb/remotes/docker/authorizer.go#L332-L344, but did not notice the difference.

**Comment by l2dy** [comment]
> > @fahedouch Should this fix be implemented in nerdctl or containerd's `remotes/docker/authorizer.go`? I will proceed testing this change after your decision.
> 
> login is on nerdctl side. SGTM to keep it on nerdctl

I missed that login is not where this issue appears, it probably has to be done in authorizer.go, which is also used in [resolver.go](https://github.com/containerd/containerd/blob/ba243ffaaf47f279e18aca980f90f9b152b257d4/remotes/docker/resolver.go#L595-L612) where the retry happens. Given that it's not CLI code anymore, I will need to make the logic more robust.

Sorry for submitting an early-stage unverified PR, I was eager to get some feedback.

**Comment by fahedouch** [comment]
> > > @fahedouch Should this fix be implemented in nerdctl or containerd's `remotes/docker/authorizer.go`? I will proceed testing this change after your decision.
> > 
> > 
> > login is on nerdctl side. SGTM to keep it on nerdctl
> 
> I missed that login is not where this issue appears, it probably has to be done in authorizer.go, which is also used in [resolver.go](https://github.com/containerd/containerd/blob/ba243ffaaf47f279e18aca980f90f9b152b257d4/remotes/docker/resolver.go#L595-L612) where the retry happens. Given that it's not CLI code anymore, I will need to make the logic more robust.
> 
> Sorry for submitting an early-stage unverified PR, I was eager to get some feedback.

login flow and corner case should be kept in the nerdctl side, the harbor auth corner case abstract for containerd.

**Pull request closed without merging**